### PR TITLE
Use iterator

### PR DESCRIPTION
**Pull request by trasher**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes

### CODE REVIEW

File: inc/olalevel.class.php [hunk 1]
```diff
@@ -300,27 +304,39 @@ static function getFirstOlaLevel($olas_id) {
...
+            'WHERE'  => [
+               'olas_id'         => $olas_id,
+               'id'              => ['<>', $olalevels_id],
+               'execution_time'  => ['>', $execution_time],
+               'is_active'       => 1
+            ],
+            'ORDER'  => 'execution_time ASC',
+            'LIMIT'  => 1
+         ]);
+         $query = "SELECT `id`
```
Comment (orthagh) [inline inc/olalevel.class.php]: remove old query

File: inc/reservation.class.php [hunk 2]
```diff
@@ -1257,24 +1293,42 @@ static function showForUser($ID) {
...
+         ],
+         'FROM'      => 'glpi_reservations',
+         'LEFT JOIN' => [
+            'glpi_reservationitems' => [
+               'ON' => [
+                  'glpi_reservationitems' => 'id',
+                  'glpi_reservations'     => 'reservationitems_id'
+               ]
+            ],
+            'glpi_entities'         => [
```
Comment (orthagh) [inline inc/reservation.class.php]: missing ON ?

File: inc/contact.class.php [hunk 3]
```diff
@@ -85,24 +85,33 @@ function defineTabs($options = []) {
...
+               ]
+            ]
+         ],
+         'WHERE'        => ['contacts_id' => $this->fields['id']]
+      ]);
+
+      if ($data = $iterator->next()) {
+         return $data;
       }
-      return "";
```
Comment (cedric-anne) [inline inc/contact.class.php]: I think this empty return should be kept.

File: inc/contractcost.class.php [hunk 4]
```diff
@@ -239,12 +239,12 @@ function initBasedOnPrevious() {
...
-                ORDER BY 'end_date' DESC, `id` DESC";
-
-      if ($result = $DB->query($query)) {
+      $iterator = $DB->request([
+         'FROM'   => $this->getTable(),
+         'WHERE'  => ['contracts_id' => $contracts_id],
+         'ORDER'  => ['end_date DESC', 'id DESC']
+      ]);
+      if ($result = $iterator->next()) {
          return $DB->fetch_assoc($result);
```
Comment (cedric-anne) [inline inc/contractcost.class.php]:
```suggestion
          return $data;
```

File: inc/link.class.php [hunk 5]
```diff
@@ -458,26 +458,33 @@ static function showForItem(CommonDBTM $item, $withtemplate = 0) {
...
-                       `glpi_links`.`name` AS name ,
-                       `glpi_links`.`data` AS data,
-                       `glpi_links`.`open_window` AS open_window
-                FROM `glpi_links`
-                INNER JOIN `glpi_links_itemtypes`
-                     ON `glpi_links`.`id` = `glpi_links_itemtypes`.`links_id`
-                WHERE `glpi_links_itemtypes`.`itemtype`='".$item->getType()."' " .
-                      getEntitiesRestrictRequest(" AND", "glpi_links", "entities_id",
-                                                 $restrict, true)."
-                ORDER BY name";
```
Comment (cedric-anne) [inline inc/link.class.php]: This order clause should be present in new query.

File: inc/calendarsegment.class.php [hunk 6]
```diff
@@ -331,12 +330,18 @@ static function showForCalendar(Calendar $calendar) {
       $canedit = $calendar->can($ID, UPDATE);
       $rand    = mt_rand();
 
-      $query = "SELECT *
-                FROM `glpi_calendarsegments`
-                WHERE `calendars_id` = '$ID'
-                ORDER BY `day`, `begin`, `end`";
-      $result = $DB->query($query);
-      $numrows = $DB->numrows($result);
+      $iterator = $Db->request([
```
Comment (cedric-anne) [inline inc/calendarsegment.class.php]:
```suggestion
      $iterator = $DB->request([
```

File: inc/event.class.php [hunk 7]
```diff
@@ -274,13 +273,13 @@ static function showForUser($user = "") {
...
       echo "<th width='60%'>".__('Message')."</th></tr>";
 
-      while ($i < $number) {
-         $ID       = $DB->result($result, $i, "id");
-         $items_id = $DB->result($result, $i, "items_id");
-         $type     = $DB->result($result, $i, "type");
-         $date     = $DB->result($result, $i, "date");
-         $service  = $DB->result($result, $i, "service");
-         $message  = $DB->result($result, $i, "message");
+      while ($data = $iterator-next()) {
```
Comment (cedric-anne) [inline inc/event.class.php]:
```suggestion
      while ($data = $iterator->next()) {
```

File: inc/itilcategory.class.php [hunk 8]
```diff
@@ -303,58 +303,61 @@ static function showForTicketTemplate(TicketTemplate $tt, $withtemplate = 0) {
...
+            ]
+         ],
+         'ORDER'  => 'name'
+      ]);
+
+      echo "<table class='tab_cadre_fixe'>";
+      echo "<tr><th colspan='3'>";
+      $itilcategory_type = $itilcategory->getType();
+      echo "<a href='".$itilcategory_type::getSearchURL()."'>";
+      echo self::getTypeName($DB->numrows($result));
```
Comment (cedric-anne) [inline inc/itilcategory.class.php]:
```suggestion
      echo self::getTypeName(count($iterator));
```

File: inc/reservation.class.php [hunk 9]
```diff
@@ -1059,12 +1073,14 @@ static function showForItem(CommonDBTM $item, $withtemplate = 0) {
...
+         $iterator = $DB->request([
+            'FROM'   => 'glpi_reservations',
+            'WHERE'  => [
+               'end'                   => ['>', $now],
+               'reservationitems_id'   => $ri->fields['id']
+            ],
+            'ORDER'  => 'begin'
+         ]);
 
          echo "<table class='tab_cadre_fixehov'><tr><th colspan='5'>";
```
Comment (cedric-anne) [inline inc/reservation.class.php]: `count($result)` is invalid in following line

File: inc/calendarsegment.class.php [hunk 10]
```diff
@@ -247,17 +251,15 @@ static function getFirstWorkingHour($calendars_id, $day) {
...
-      return false;
+      $result = $DB->request([
+         'SELECT' => ['MIN' => 'begin AS minb'],
+         'FROM'   => 'glpi_calendarsegments',
+         'WHERE'  => [
+            'calendars_id' => $calendars_id,
+            'day'          => $day
+         ]
+      ])->next();
+      return $result['minb'];
```
[the remark this answers is not in the document]
Comment (trasher) [reply inc/calendarsegment.class.php]: Looking at usage.. It iw waiting for a string :D I guess new is not worse than old

**Pull request merged**